packages/runtime-dom/src/directives/vModel.ts

```diff
@@ -86,9 +86,10 @@ export const vModelText: ModelDirective<
     el[assignKey] = getModelAssigner(vnode)
     // avoid clearing unresolved text. #2302
     if ((el as any).composing) return
-
     const elValue =
-      number || el.type === 'number' ? looseToNumber(el.value) : el.value
+      (number || el.type === 'number') && !/^0\d/.test(el.value)
+        ? looseToNumber(el.value)
```

Comment:
I made two small tweaks here:
1. This leading 0 check should apply to both type="number" and .number modifier cases
2. The leading 0 check should not apply to decimals like `0.123`.